### ISSUE TITLE
Clean way to update repeatable job data?

### ISSUE THREAD

**Issue by gmcnaught** [OptimalBits/bull#742]
The use case would be to update the data for the next run of a job in an atomic fashion, which currently seems to only be possible by using removeRepeatableJob and re-adding the repeatable job, and only while the job is not actively running.

**Comment by manast** [comment]
I will mark this issue as a future enhancement.

**Comment by lossau** [comment]
I also need this.
Has anyone found a way to update a repeatable job's data during processing?

**Comment by abelosorio** [comment]
This is a really needed feature.

The documentation was not clear to me, and I was adding the job every time my node runs causing multiple jobs to run in every `npm start`.

I cannot use the @gmcnaught workaround because I don't allways know what is the old value for `RepeatOpts`. So I'm cleanning the queue before adding it in again, and using a `jobId` (which I have defined in a file of constants `jobs_constants`) to avoid duplicated queues:

```javascript
import Queue from 'bull';

import { REMINDER_FOR_TRACKING_HOURS } from '../config/jobs_constants';

module.exports = (RedisUrl, RedisOptions) => {
  // Here is where I build the `queue` object and write all the jobs specifications.

  queue.empty();
  queue.add(null, {
    jobId: REMINDER_FOR_TRACKING_HOURS,
    repeat: { cron: '0 10 26-31 * *' }
  });
}
```

A `addOrUpdate` method would be really useful. Maybe with required `jobId` or some unique identifier.

**Comment by manast** [comment]
I will look into this as soon as I have a free slot.

**Comment by mugli** [comment]
Any update on this?

**Comment by kizdolf** [comment]
Having a `job.update()` working for repeatable would be really nice. I expose jobs to a graphQL but without any data in them ... :/

**Comment by LukaszWiktor** [comment]
I also need to update repeatable jobs.

**Comment by gmcnaught** [comment]
I think I've mentioned this in another thread:

We augmented Bull by storing job Data for repeatable jobs in a hash by jobName, that way we can atomically update the hash with the new job data, and then the first step in job.process pulls the hash data and updates the current job, so future jobs will have this data. 

We've explored adding this in to the redis/lua automation to integrate it into the system directly and to reduce the number of network calls, but its not been a high priority.
We have also considered handling delete in the same way - creating a queue of repeatable jobs that need to be deleted, and asynchronously deleting them on their next scheduled run.

**Comment by LukaszWiktor** [comment]
Thanks @gmcnaught!

I was trying to keep all the required information to run a job in its data, but it turns out I need to store it somewhere on the side.

**Comment by sor-barroso** [comment]
I've faced the same need. It would be awesome if we had an easiest way to perform updates on repeatable jobs.

**Comment by crispinkoech** [comment]
Any update on this? Does job.update() work for repeatable jobs as well?

**Comment by wasd171** [comment]
This feature would be very useful for the cron jobs where we need to query something since the last task invocation, it's a pity that this doesn't work by default

**Comment by grapevineai** [comment]
@manast Will this be implemented in BullMQ as well?

**Comment by MaximilianHollis** [comment]
Any progress on this issue?

**Comment by hf29h8sh321** [comment]
> Any progress on this issue?

No

**Comment by RanKey1496** [comment]
5 years since someone asked for it, it would be great to do it

**Comment by paco-sisutech** [comment]
In case anyone would find it usable:
`
export async function updateJobsOfQueueWithNewPattern(queue: Queue, pattern: string | undefined) {
    if (pattern) {
        const connection = await queue.client;
        const key = (await connection.keys(`bull:${queue.name}:repeat:*:*`))[0];
        const redisData = await connection.hgetall(key);
        if (redisData) {
            const jobs = (await queue.getRepeatableJobs()).filter((job) => job.pattern !== pattern);
            for (const job of jobs) {
                log.info(`Update job ${job.name} with new pattern ${pattern}`);
                const options = { repeat: { pattern }};
                await queue.removeRepeatableByKey(job.key);
                await addJob(job.name, queue, JSON.parse(redisData.data), options);
            }
        }
    }
}
`
A bit dirty but does what I wanted

**Comment by dlebee** [comment]
Also would love this feature :)

**Comment by Pandaver** [comment]
It has been 7 years and this problem has not been solved. I want to update it by deleting and then creating, but I cannot obtain the old data through getRepeatableJobs. Please update it @manast

**Comment by annieTechno** [comment]
Hey @manast 
Can you please provide an ETA when this feature is expected? or if it is not, please provide a reason why recreating a job is the only solution provided, it is very inefficient.

**Comment by PGabriel20** [comment]
After seven years we still don't have a way to update a repeatable job without this workaround of deleting and adding it again, also a method to get a repeatable job by id would be much appreciated and would facilitate doing this workaround.

**Comment by khiet-vo** [comment]
So this https://docs.bullmq.io/guide/jobs/repeatable#updating-repeatable-jobs-options
Is just in theory

**Comment by roggervalf** [comment]
yes, that's currently possible using bullmq but also we are recommending to use https://docs.bullmq.io/guide/job-schedulers
